velox/type/Filter.cpp

```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+}
+
+folly::dynamic HugeintRange::serialize() const {
+  auto obj = Filter::serializeBase("HugeintRange");
+  char lower[sizeof(int128_t)];
+  HugeInt::serialize(lower_, lower);
+  obj["lower"] = std::string(lower);
+  char upper[sizeof(int128_t)];
+  HugeInt::serialize(upper_, upper);
+  obj["upper"] = std::string(upper);
+  return obj;
+}
+
+FilterPtr HugeintRange::create(const folly::dynamic& obj) {
+  auto lower = HugeInt::deserialize(obj["lower"].asString().c_str());
+  auto upper = HugeInt::deserialize(obj["upper"].asString().c_str());
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  return std::make_unique<HugeintRange>(lower, upper, nullAllowed);
+}
+
+bool HugeintRange::testingEquals(const Filter& other) const {
+  auto otherHugeintRange = dynamic_cast<const HugeintRange*>(&other);
+  return otherHugeintRange != nullptr && Filter::testingEquals(other) &&
+      lower_ == otherHugeintRange->lower_ &&
+      upper_ == otherHugeintRange->upper_;
+}
+
+folly::dynamic BigintValuesUsingHashTable::serialize() const {
+  auto obj = Filter::serializeBase("BigintValuesUsingHashTable");
+  obj["min"] = min_;
+  obj["max"] = max_;
+
+  folly::dynamic values = folly::dynamic::array;
+  for (auto v : values_) {
+    values.push_back(v);
+  }
+  obj["values"] = values;
+
+  folly::dynamic hashTable = folly::dynamic::array;
+  for (auto h : hashTable_) {
+    hashTable.push_back(h);
+  }
+  obj["hashTable"] = hashTable;
+
+  return obj;
+}
+
+FilterPtr BigintValuesUsingHashTable::create(const folly::dynamic& obj) {
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  auto min = obj["min"].asInt();
+  auto max = obj["max"].asInt();
+
+  std::vector<int64_t> values;
+  auto valArr = obj["values"];
+  values.reserve(valArr.size());
+  for (const auto& v : valArr) {
+    values.push_back(v.asInt());
+  }
+
+  std::vector<int64_t> hashTable;
+  auto hashArr = obj["hashTable"];
+  hashTable.reserve(hashArr.size());
+  for (const auto& h : hashArr) {
+    hashTable.push_back(h.asInt());
+  }
+  auto res = std::make_unique<BigintValuesUsingHashTable>(
+      min, max, values, nullAllowed);
+
+  return res;
+}
+
+bool BigintValuesUsingHashTable::testingEquals(const Filter& other) const {
+  auto otherBigintValuesUsingHashTable =
+      dynamic_cast<const BigintValuesUsingHashTable*>(&other);
+  bool res = otherBigintValuesUsingHashTable != nullptr &&
+      Filter::testingEquals(other) &&
+      min_ == otherBigintValuesUsingHashTable->min_ &&
+      max_ == otherBigintValuesUsingHashTable->max_ &&
+      containsEmptyMarker_ ==
+          otherBigintValuesUsingHashTable->containsEmptyMarker_ &&
+      sizeMask_ == otherBigintValuesUsingHashTable->sizeMask_ &&
+      hashTable_.size() == otherBigintValuesUsingHashTable->hashTable_.size() &&
+      values_.size() == otherBigintValuesUsingHashTable->values_.size();
+
+  if (!res) {
+    return false;
+  }
+
+  for (size_t i = 0; i < values_.size(); ++i) {
```

Comment:
nit: perhaps, add a comment to explain that values_ are sorted, hence, can be compared pair-wise